### PR TITLE
[REPO-5239] - Added filter by displayName to /groups endpoint

### PR DESCRIPTION
**Pull request by Giovanni007**
Add tha ability to filter groups by displayName when using /groups endpoint and the "where"query parameter. 

e.g `where=(displayName in ('AGroupName'))`

At the moment it is possible specify only 1 group name in the list of displayNames ( the same as for zones).

I decided to avoid to use the pre-existing filter present in  authorityService.getAuthoritiesInfo() for the following motivation:

- I tested it and it allows wildcards

-  It filters by groupName instead of groupDisplayName

### CODE REVIEW

File: src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java [hunk 1]
```diff
@@ -262,6 +270,25 @@ else if (zonesParam.isEmpty())
...
+        if (displayNamesParam.size() > MAX_DISPLAY_NAMES)
+        {
+            throw new IllegalArgumentException("A maximum of " + MAX_DISPLAY_NAMES + " displayNames may be specified.");
+        }
+        else if (displayNamesParam.isEmpty())
+        {
+            throw new IllegalArgumentException("DisplayNames filter list cannot be empty.");
+        }
+
+        displayNamesParam.forEach(groupName -> {
```
Comment (jesty) [inline src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java:284]: can you use a filter method instead of the foreach?

File: src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java [hunk 2]
```diff
@@ -213,7 +221,7 @@ public Group getGroup(String groupId, Parameters parameters) throws EntityNotFou
         PagingResults<AuthorityInfo> pagingResult;
         try
         {
-            pagingResult = getAuthoritiesInfo(authorityType, isRootParam, zoneFilter, rootAuthorities, sortProp, paging);
+            pagingResult = getAuthoritiesInfo(authorityType, isRootParam, zoneFilter, displayNameFilter, rootAuthorities, sortProp, paging);
```
Comment (jesty) [inline src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java:224]: This method has too many parameters, please refactoring it, at least using a parameter object.

File: src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java [hunk 3]
```diff
@@ -367,7 +395,7 @@ else if (zonesParam.isEmpty())
                 // Get authorities using canned query but without using
                 // the requested paginating now because we need to filter out
                 // the root authorities.
-                PagingResults<AuthorityInfo> nonPagingResult = authorityService.getAuthoritiesInfo(authorityType, zoneFilter, null, sortProp.getFirst(), sortProp.getSecond(),
+                PagingResults<AuthorityInfo> nonPagingResult = authorityService.getAuthoritiesInfo(authorityType, zoneFilter, groupNamesFilter, sortProp.getFirst(), sortProp.getSecond(),
```
Comment (mauriziovitale) [inline src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java:398]: I think we are missing the null as authority

Comment (Giovanni007) [reply src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java]: I use that value to pass diplayName to the canned query

File: src/test/java/org/alfresco/rest/api/tests/GroupsTest.java [hunk 4]
```diff
@@ -1856,4 +1857,105 @@ private Group generateSubGroup(Group parentGroup)
 
         return subGroup;
     }
+
+    private void testGetGroupsWithDisplayNameFilter() throws Exception
```
Comment (gavincornwell) [inline src/test/java/org/alfresco/rest/api/tests/GroupsTest.java:1861]: Does the authorityService.getAuthoritiesInfo() method accept wildcards for display name? If so, we should have a test to ensure multiple results are returned.

Comment (Giovanni007) [reply src/test/java/org/alfresco/rest/api/tests/GroupsTest.java:1861]: At the moment my code is not consisent, it  doesn't support wildcards in the case the parameter isRoot=true because in this specific case we don't use the canned query but the already loaded list of root authorities. 
I think it's better I change that and I avoid to use getAuthoritiesInfo to filter by displayName . It accepts wildcards (we don't need them) and if I want to support them I'll have to duplicate all the logic we have in [GetAuthoritiesCannedQuery.java](https://github.com/Alfresco/alfresco-repository/blob/5e247e50b48eeda597eefada2cc990b1797e3689/src/main/java/org/alfresco/repo/security/authority/GetAuthoritiesCannedQuery.java).
If use getAuthoritiesInfo() with displayNamefilter null as it was, I'll have to filter the result before the pagination. Is this a better solution or should I find a way to support wildcards when using the already loaded list of root authorities?

Comment (gavincornwell) [reply src/test/java/org/alfresco/rest/api/tests/GroupsTest.java:1861]: For consistency we need to support display name filtering for root groups if we support it for non-root groups.

Looking at the JavaDoc for the getAuthoritiesInfo method (https://github.com/Alfresco/alfresco-repository/blob/ac38ac94ff4f9cbdf2671a9517781bda389a13c4/src/main/java/org/alfresco/service/cmr/security/AuthorityService.java#L169) it implies the name filter is only a startsWith case insensitive check (there are no wildcard characters passed) which should be fairly easy to replicate without having to duplicate the logic in GetAuthoritiesCannedQuery. 

In other words could you just modify the logic in displayNamePredicate here: https://github.com/Alfresco/alfresco-remote-api/blob/1f38ad40cf884a201755263ee93bcfcfa0872153/src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java#L499 ?

Comment (Giovanni007) [reply src/test/java/org/alfresco/rest/api/tests/GroupsTest.java:1861]: I tested it and it allows wildcards (eg. ?, .*), but the main problem is that it uses both the displayName and the authName to filter the result. I decided to not use it at the moment, the name filter will be a basic case insensitive comparation.

Comment (gavincornwell) [reply src/test/java/org/alfresco/rest/api/tests/GroupsTest.java:1861]: OK
